src/main/scala/cognite/spark/v1/AssetsHierarchyBuilder.scala

```diff
@@ -0,0 +1,223 @@
+          client.assets.updateByExternalId(
+            Seq(
+              root.externalId -> AssetsIngestSchema
+                .toAssetUpdate(root)
+                .copy(parentExternalId = None)).toMap) *> IO.unit
+        }
+    }
+
+  def nodesToDeleteInsertUpdate(
+      root: AssetsIngestSchema,
+      leaves: Set[AssetsIngestSchema],
+      cdfTree: Seq[Asset]): (Seq[Asset], Set[AssetsIngestSchema], Set[AssetsIngestSchema]) = {
+    // Insert assets that are not present in CDF
+    val toInsert = leaves.filter(a => !cdfTree.map(_.externalId.get).contains(a.externalId))
+
+    // Keep assets that are in both source and cdf, delete those only in CDF
+    val (toKeep, toDelete) =
+      cdfTree.partition(a => (Seq(root) ++ leaves).map(_.externalId).contains(a.externalId.get))
+
+    // Filter out assets that would not be changed in an update
+    val toUpdate = leaves.filter(a => needsUpdate(a, toKeep))
+
+    (toDelete, toInsert, toUpdate)
+  }
+
+  def needsUpdate(leaf: AssetsIngestSchema, cdfTree: Seq[Asset]): Boolean = {
+    // Find the matching Asset in CDF
+    val cdfAsset = cdfTree.find(a => a.externalId.contains(leaf.externalId))
+    cdfAsset match {
+      case Some(asset) =>
+        // Find the parent Asset in CDF to be able to get the externalId of the parent
+        cdfTree.find(a => asset.parentId.contains(a.id)) match {
+          case Some(parentAsset) =>
+            !isMostlyEqual(leaf, asset) ||
+              !parentAsset.externalId.contains(leaf.parentExternalId)
+          case None =>
+            false
+        }
+      case None => false
+    }
+  }
+
+  def isMostlyEqual(leaf: AssetsIngestSchema, asset: Asset): Boolean =
+    leaf.description == asset.description &&
+      leaf.metadata == asset.metadata &&
+      leaf.name == asset.name &&
+      leaf.source == asset.source
+
+  def validateAndOrderInput(
+      tree: Array[AssetsIngestSchema]): (AssetsIngestSchema, Set[AssetsIngestSchema]) = {
+    val (root, leaves) = splitIntoRootAndLeaves(tree)
+
+    (root, getValidatedTree(root, leaves))
+  }
+
+  def splitIntoRootAndLeaves(
+      tree: Array[AssetsIngestSchema]): (AssetsIngestSchema, Array[AssetsIngestSchema]) = {
+    val (maybeRoots, leaves) = tree.partition(_.parentExternalId == "")
+
+    val root = maybeRoots.toList match {
+      case singleRoot :: Nil => singleRoot
+      case Nil => throw NoRootException()
+      case _ => throw MultipleRootsException()
+    }
+    (root, leaves)
+  }
+
+  def getValidatedTree(
+      root: AssetsIngestSchema,
+      leaves: Array[AssetsIngestSchema]): Set[AssetsIngestSchema] = {
+    val visited = Set[AssetsIngestSchema]()
+    val insertableTree = iterateChildren(leaves, Array(root.externalId), visited)
+
+    if (insertableTree.map(_.externalId) == leaves.map(_.externalId).toSet) {
+      insertableTree
+    } else {
+      val assetsNotInTree =
+        leaves.filter(a => insertableTree.map(_.externalId).contains(a.externalId)).map(_.externalId)
```

Comment:
`leaves.map(_.externalId).filter(externalId => insertableTree.map(_.externalId).contains(externalId))`
but since we don't want to have to do this mapping on insertableTree for ever element of leaves we should do that once at the top: `val insertableTreeExternalIds = insertableTree.map(_.externalId)`